vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -1,1011 +1,508 @@
-
-        # Discover hosts and virtual machines
-        server_instance = self._get_server_instance(instance)
-        use_guest_hostname = is_affirmative(instance.get("use_guest_hostname", False))
-        all_objs = self._get_all_objs(
-            server_instance, regexes, include_only_marked, tags, use_guest_hostname=use_guest_hostname
+    def refresh_metrics_metadata_cache(self):
+        """Request the list of counters (metrics) from vSphere and store them in a cache."""
+        self.log.info(
+            "Refreshing the metrics metadata cache. Collecting all counters metadata for collection_level=%d",
+            self.collection_level,
         )
-
-        self.mor_objects_queue.fill(i_key, dict(all_objs))
-        self.cache_config.set_last(CacheConfig.Morlist, i_key, time.time())
-
-    @trace_method
-    def _process_mor_objects_queue_async(self, instance, mors):
-        """
-        Process a batch of items popped from the objects queue by querying the available
-        metrics for these MORs and then putting them in the Mor cache
-        """
-        t = time.time()
-        i_key = self._instance_key(instance)
-        server_instance = self._get_server_instance(instance)
-        perfManager = server_instance.content.perfManager
-
-        # For non realtime metrics, we need to specifically ask which counters are available for which entity,
-        # so we call perfManager.QueryAvailablePerfMetric for each cluster, datacenter, datastore
-        # This should be okay since the number of such entities shouldn't be excessively large
-        for mor in mors:
-            mor_name = str(mor['mor'])
-            available_metrics = {m.counterId for m in perfManager.QueryAvailablePerfMetric(entity=mor["mor"])}
-            try:
-                self.mor_cache.set_metrics(i_key, mor_name, self._compute_needed_metrics(instance, available_metrics))
-            except MorNotFoundError:
-                self.log.error("Object '%s' is missing from the cache, skipping. ", ensure_unicode(mor_name))
+        t0 = Timer()
+        counters = self.api.get_perf_counter_by_level(self.collection_level)
+        self.gauge("datadog.vsphere.refresh_metrics_metadata_cache.time", t0.total(), tags=self.base_tags, raw=True)
+        self.log.info("Collected %d counters metadata in %.3f seconds.", len(counters), t0.total())
+
+        for mor_type in self.collected_resource_types:
+            allowed_counters = []
+            for c in counters:
+                metric_name = format_metric_name(c)
+                if metric_name in ALLOWED_METRICS_FOR_MOR[mor_type] and not is_metric_excluded_by_filters(
+                    metric_name, mor_type, self.metric_filters
+                ):
+                    allowed_counters.append(c)
+            metadata = {c.key: format_metric_name(c) for c in allowed_counters}
+            self.metrics_metadata_cache.set_metadata(mor_type, metadata)
```

Comment:
No `with ...` needed here?

Reply:
All this function is run within a `with cache.update()` block already. This is clear when looking at the `check()` method